README.md
# kicad_wiretools
Wire toolset for KiCad

* Shielding tool for creating web of shielding

## Code of conduct

New features should be created in new branch, so that 'master' branch is always stable. When new features are tested, they can be joined to master branch.

## TODO

** Wire & simulation tools

* [ ] Automatic naming of nets
* [ ] Frequency behaviour for each net; Resistance, impedance, ...
* [ ] Damping per net
* [ ] Trace-trace capacitance -> damping
* [ ] Shielding efficiency -> dB / Hz
* [ ] EMI radiance
* [ ] Digital pulse response per net
* [ ] Capacitance between net (trace) and shielding
* [ ] Graphically select traces
  * cut traces
  * delete traces
  * rename nets
* [ ] Calculations
  * [ ] Voltage drop
  * [x] Trace length per net
  * [ ] Trace inductance per net
  * [ ] Trace impedance
  * [ ] Power loss per trace
  * [ ] IPC 2221 maximum current per trace I = K * dT^0.44 * (W*H)^0.725; K=[0.024, 0.048]
  * [ ] Trace comparisation
  
** Shielding

  * [x] Free angle selection for hash
  * [x] Calculate the procentual coverage of shielding



Changelog
0.2.4 - 2018-11-28
------------------

* Tracelength calculation
* Trace maximum current calculation according to IPC 2223
* Bugfixes

0.2.3 - 2018-10-31
------------------

* Added shielding coverage calculation

0.2.2 - 2018-10-30
------------------

* Added angle control to hash generation

0.2.1 - 2018-09-23
------------------

* bugfixes

0.2 - 2018-09-23
----------------

* Basic functionality ready



tests/test_format.py
#!/usr/bin/python

import unittest

from traceinfo import format_number

data = (
    (1, "1.000"),
    (-1, "-1.000"),
    (10, "10.000"),
    (100, "100.000"),
    (-100, "-100.000"),
    (1e3, "1.000k"),
    (1e4, "10.000k"),
    (1e5, "100.000k"),
    (-1e5, "-100.000k"),
    (1e6, "1.000M"),
    (1e-1, "100.000m"),
    (1e-2, "10.000m"),
    (1e-3, "1.000m"),
    (1e-4, "100.000u"),
    (1e-5, "10.000u"),
    (1e-6, "1.000u"),
    )


class TestFormatFunc(unittest.TestCase):
    def test_format(self):
        for inp, out in data:
            print "input of '%s' should be '%s'" %(inp, out)
            self.assertEqual(format_number(inp), out)


tests/test_traceinfo.py
import unittest

from traceinfo import calculate_resistance, calculate_length

class Point(object):
    def __init__(self, **args):
        self.x = args['x'] or 0
        self.y = args['y'] or 0
    def y(self, yy):
        self.y = yy
        return self.y
    def x(self, xx):
        self.x = xx
        return self.x
    
    

testcases_resistance = (
    (Point(x=0, y=0), Point(x=10, y=0), 1e-3, 35e-6, 1.72e-8, 0.00491429),
)

testcases_length = (
    (Point(x=0, y=0), Point(x=10e+5, y=0), 1),
)

class TestTraceInfo(unittest.TestCase):
    def test_calculate_resistance(self):
        for start, end, width, thick, rho, expected in testcases_resistance:
            self.assertEqual(calculate_resistance(start, end, width, rho), expected)


    def test_calculate_length(self):
        for start, end, expected in testcases_length:
            self.assertEqual(calculate_length(start, end), expected)
if __name__ == '__main__':
    unittest.main()


__init__.py
import pcbnew

__version__ = "0.2.4"

print("Initializing kicad_wiretools version {}".format(__version__))

#import wiretools

from .shielding import HashShieldGenerator

#import module_loader

#import wiretools_dumper

# -----------------------------------------------------------
print("Register shielding tools")
HashShieldGenerator().register()
print(" HashShieldGenerator registration compeleted.")
# -----------------------------------------------------------

#print("Register Wiretools")
#wiretools.WireTools().register()

#print("done adding kicad_wiretools")

# -----------------------------------------------------------
print("Register TraceInfo tools")

from .traceinfo import TraceInfoGenerator
TraceInfoGenerator().register()

print(" TraceInfoGenerator registration completed.")
# -----------------------------------------------------------


wiretools.py
import pcbnew

import inspect
import os

filename = inspect.getframeinfo(inspect.currentframe()).filename
path = os.path.dirname(os.path.abspath(filename))

board = pcbnew.GetBoard()

SCALE = 1000000

if hasattr(pcbnew, "LAYER_ID_COUNT"):
    pcbnew.PCB_LAYER_ID_COUNT = pcbnew.LAYER_ID_COUNT

if not board:
    Exception("Unable to get board handle")
    
def coordsFromPolySet(ps):
    str = ps.Format()
    lines = str.split('\n')
    numpts = int(lines[2])

    # there is extra two \n in each line
    pts = [[int(n) for n in x.split(" ")] for x in lines[3:-2]]
    return pts

"""Get all pads for a NET"""
def padsForNet(net):
    retval = []
    for pad in board.getPads():
        # first get the netinfo, then get the netcode (int)
        if pad.GetNet().GetNet() == net:
            retval.append(pad)
    return retval

# from collections import defaultdict

#ds = board.GetDesignSettings()

def pan_and_zoom(x, y, width, height):
    xx = pcbnew.FromMM(x)
    yy = pcbnew.FromMM(y)
    ww = pcbnew.FromMM(width)
    hh = pcbnew.FromMM(height)
    pcbnew.WindowZoom(xx, yy, ww, hh)

def dump_board_boundingbox():
    board = pcbnew.GetBoard()
    boardbox = board.ComputeBoundingBox()
    boardx1 = boardbox.GetX()
    boardy1 = boardbox.GetY()
    boardwidth = boardbox.GetWidth()
    boardheight = boardbox.GetHeight()
    print("this board is at position {},{} {} widde and {} high".format(boardx1, boardy1, boardwidth, boardheight))
    return (boardx1, boardy1, boardwidth, boardheight)
    
def get_layertable():
    board = pcbnew.GetBoard()
    layertable = {}
    numlayers = pcbnew.LAYER_ID_COUNT
    for i in range(numlayers):
        layertable[i] = board.GetLayerName(i)
        print("{} {}".format(i, board.GetLayerName(i)))
    return layertable
    
def dump_net_tracks(net):
    board = pcbnew.GetBoard()
    clktracks = board.TracksInNet(clknet.GetNet())
    for track in clktracks:
        print("{},{} -> {},{} width {} layer {}".format(track.GetStart().x/SCALE, track.GetStart().y/SCALE, track.GetEnd().x/SCALE, track.GetEnd().y/SCALE, track.GetWidth()/SCALE, layertable[track.GetLayer()]))

#tracks = board.GetTracks()
#for track in tracks:
#    print("track: {}".format(track))


class Wiretools(pcbnew.ActionPlugin):
    def defaults(self):
        self.name = "Wiretools utility"
        self.category = "Show Info from PCB"
        self.description = "Generate information from PCB"

    def _prepare_report(self):
        self.stream = None

    def Run(self):
        self._board = pcbnew.GetBoard()
        


spikes/gdemo.py
#!/usr/bin/env python

from math import sin, cos, pi, fabs, tan, sqrt
from graphics import *
    

def draw_hash(screen, angle, left, top, right, bottom, offset_left=5, offset_right=5, offset_top=5, offset_bottom=5):
    x = left
    y = top

    width = 5
    
    real_left = left + offset_left
    real_right = right - offset_right
    real_top = top + offset_top
    real_bottom = bottom - offset_bottom
    
    line = Line(Point(real_left, real_top),
                Point(real_right, real_top))
    line.setOutline('red')
    line.setWidth(width)
    line.draw(screen)

    line = Line(Point(real_left, real_bottom),
                Point(real_right, real_bottom))
    line.setOutline('red')
    line.setWidth(width)
    line.draw(screen)

    line = Line(Point(real_left, real_top),
                Point(real_left, real_bottom))
    line.setOutline('red')
    line.setWidth(width)
    line.draw(screen)

    line = Line(Point(real_right, real_top),
                Point(real_right, real_bottom))
    line.setOutline('red')
    line.setWidth(width)
    line.draw(screen)
    

    alfa = angle*pi/180.0
    

    pitch = 20
    nn = sqrt(pow(real_bottom - real_top, 2) +
              pow(real_right - real_left, 2)) / (pitch/cos(alfa)) - 1

    n = 0
    xx = 0
    yy = 0

    my = real_bottom - real_top
    mx = real_right - real_left

    dy = pitch / cos(alfa)

    print "pitch({}, -> {}) => n={}".format(pitch, dy, nn)
    
    while n < nn:
        n += 1

        zy = n * pitch / cos(alfa)
        zx = zy / tan(alfa)       

        ax0 = real_left
        ax1 = zx + real_left
        ay0 = real_top + zy
        ay1 = real_top

        if zy > my:
            ax0 += (zy-my) / tan(alfa)
            ay0 = real_bottom
        if zx > mx:
            ay1 = real_top + ( (zx - mx) * tan(alfa) )
            ax1 = real_right            


        qx = n * pitch / cos(alfa)
        qy = qx / tan(alfa)


        bx0 = real_left
        bx1 = qx+real_left
        by0 = real_bottom-qy
        by1 = real_bottom

        if qy > my:
            by0 = real_top
            bx0 += (qy-my) * tan(alfa)
        if qx > mx:
            bx1 = real_right
            by1 -= ((qx-mx) / tan(alfa))
        print "Got[{}]: ({}, {}, -> {}, {}) zy={}, zx={}".format(n, ax0, ay0, ax1, ay1, zy, zx)
        line = Line(Point(ax0, ay0), Point(ax1, ay1))
        line.setOutline('red')
        line.setWidth(width)
        line.draw(screen)

        print "Got[{}]: ({}, {}, -> {}, {}) qy={}, qx={}".format(n, bx0, by0, bx1, by1, qy, qx)
        line = Line(Point(bx0, by0), Point(bx1, by1))
        line.setOutline('red')
        line.setWidth(width)
        line.draw(screen)



def main():
    mx1, my1, mx2, my2 = (0, 0, 640, 640)
    x1, y1, x2, y2 = (mx1+80, my1+80, mx2-80, my2-80)

    win = GraphWin('Main', mx2-mx1, my2-my1)

    pad = Rectangle(Point(x1, y1), Point(x2, y2))
    pad.setOutline('gray')
    pad.draw(win)

    oleft = 5
    oright = 5
    otop = 5
    obottom = 5

    ang = 22.5

    draw_hash(win, ang, x1, y1, x2, y2, oleft, oright, otop, obottom)

    line = Line(Point(20,20), Point(200,200))
    line.setWidth(10)
    line.setOutline('black')
    line.draw(win)
    
    # wait for mouse event
    win.getMouse()
    win.close()


if __name__ == '__main__':
    main()



traceinfo.py
# Copyright (c) 2018 Tommi Rintala, New Cable Corporation Ltd

#import sys
from math import sqrt, log, exp

import wx
import wx.aui

#from pcbnew import *
import pcbnew

# some code stolen from:
# https://github.com/KiCad/kicad-source-mirror/blob/master/pcb_calculator/tracks_width_versus_current.cpp

#ToUnits = ToMM
#FromUnits = FromMM

RHO_CU = 1.72e-8  # Copper resistivity
CU_THICK = 18e-6  # Board thickness
MAX_TEMP = 20     # maximum temperature rise (degrees Celcius)

def format_number(inputvalue, dec=3):
    """Format number more nicely"""
    value = inputvalue
    post = ""
    if abs(value) >= 1e9:
        value /= 1e9
        post = "G"
    elif abs(value) >= 1e6:
        value /= 1e6
        post = "M"
    elif abs(value) >= 1e3:
        value /= 1e3
        post = "k"
    elif abs(value) >= 0:
        if abs(value) < 1e-12:
            value *= 1e12
            post = 'p'
        elif abs(value) < 1e-9:
            value *= 1e9
            post = 'n'
        elif abs(value) < 1e-6:
            value *= 1e6
            post = 'u'
        elif abs(value) < 1e-3:
            value *= 1000
            post = 'm'


    return "%.3f%s" % (value, post)


def calculate_resistance(start, end, width, height, rho_cu=RHO_CU):
    """Calculate resistance"""
    length = calculate_length(start, end)
    resistance = (rho_cu * length) / (width * height)
    #sqrt(pow(ToMM(end.y)-ToMM(start.y), 2) + pow(ToMM(end.x)-ToMM(start.x), 2))
    #height = 35e-6
    #area = pcbnew.ToMM(width) * height * 1000
    #resistance = rho_cu * length/area
    return resistance

def calculate_inductance(areaMM2):
    return 0



def calculate_area(point1, point2, width):
    area = calculate_length(point1, point2) * pcbnew.ToMM(width)
    return area


def calculate_trace_width(current_A, thickness_m, delta_T_C, use_internal_layer):
    if use_internal_layer:
        scale = 0.024
    else:
        scale = 0.048
    dtmp = log(current_A) - log(scale) - 0.44 * log(delta_T_C) - 0.725 * log(thickness_m)
    dtmp /= 0.725
    trackwidth = exp(dtmp)
    return trackwidth

def calculate_max_current(width_m, thickness_m, delta_T_C, use_internal_layer):
    if use_internal_layer:
        scale = 0.024
    else:
        scale = 0.048
    area = thickness_m * width_m
    current = scale * pow(delta_T_C, 0.44) * pow(area, 0.725)
    return current

def calculate_length(point1, point2):
    return sqrt(pow(pcbnew.ToMM(point1.y)-pcbnew.ToMM(point2.y), 2)
                + pow(pcbnew.ToMM(point1.x)-pcbnew.ToMM(point2.x), 2))

def traceinfo(cu_thick=CU_THICK, internal_layer=True):
    resistance = {0:0}
    inductance = {0:0}
    #maxvoltage = {0:999999}
    maxcurrent = {0:999999}
    voltagedrop = {0:0}
    powerloss = {0:0}
    netnames = {0:u'default'}
    lengths = {0:0}

    for trace in pcbnew.GetBoard().GetTrackWidthList():
        print("trace width: " + str(trace))
        for item in pcbnew.GetBoard().GetTracks():
            print("track: " + str(trace))
            if isinstance(item, pcbnew.VIA):
                pos = item.GetPosition()
                drill = item.GetDrillValue()
                width = item.GetWidth()
                print(" * Via: ", (pcbnew.ToMM(pos), pcbnew.ToMM(drill), pcbnew.ToMM(width)))
            elif isinstance(item, pcbnew.TRACK):
                if item.GetNet():
                    net = item.GetNet()
                    net_id = net.GetNet()
                    net_name = net.GetNetname() or u''
                else:
                    net_id = 0
                    net_name = u'default'

                start = item.GetStart()
                end = item.GetEnd()
                width = item.GetWidth()

                area = calculate_area(start, end, width)
                length = calculate_length(start, end)
                #net_name = net.GetNetname() or u''
                print(" * TRACK: ", net_name, area)
                if net_id not in resistance:
                    resistance[net_id] = 0
                if net_id not in inductance:
                    inductance[net_id] = 0
                if net_id not in voltagedrop:
                    voltagedrop[net_id] = 0
                if net_id not in powerloss:
                    powerloss[net_id] = 0
                if net_id not in maxcurrent:
                    maxcurrent[net_id] = 99999999
                if net_id not in netnames:
                    netnames[net_id] = net_name
                if net_id not in lengths:
                    lengths[net_id] = 0
                #rr = calculate_resistance(start, end, width)
                ii = calculate_inductance(area)

                #print(" -> area= ", area, ", R= ", rr, " L= ", ii)

                cable_current = 1

                extResistance = (RHO_CU * length) / (width * cu_thick)
                external_voltage = extResistance * cable_current
                lengths[net_id] += length
                resistance[net_id] += extResistance
                voltagedrop[net_id] += external_voltage
                powerloss[net_id] += external_voltage * cable_current
                #inductance[net_id] += ii
                maxcurr = calculate_max_current(width, cu_thick, 20, internal_layer)
                maxcurrent[net_id] = min(maxcurr, maxcurrent[net_id])

    for net in resistance:
        print("Total resistance for net: ", net, " (", netnames[net], ") is ", resistance[net], " Ohm")
        print(" voltagedrop: ", voltagedrop[net])
        print(" powerloss:   ", powerloss[net])
        print(" max current: ", maxcurrent[net])
    for net in inductance:
        print("Total inductance for net: ", net, " is ", inductance[net], " Ohm")
    return (resistance, inductance, powerloss, voltagedrop, maxcurrent, netnames, lengths)


class TraceInfoGenerator(pcbnew.ActionPlugin):
    def defaults(self):
        self.name = "Generate Trace Info"
        self.category = "Info PCB"
        self.description = "This plugin gives some info about traces"

        self._board = pcbnew.GetBoard()


    def get_max_voltage(self, K_const, temprise, width, thickness):
        return K_const * pow(temprise, 0.44) * pow(width * thickness, 0.725)

    def Run(self):
        print("----------------------------------")
        print("Starting Traceinfo plugin")

        class DisplayResults(wx.Dialog):
            def __init__(self, parent):
                wx.Dialog.__init__(self, parent, id=wx.ID_ANY, title="Traceinfo Results")
                #self.SetIcon()
                panel = wx.Panel(self)

                #pcb = pcbnew.GetBoard()
                #nets = pcb.GetNets()
                res, ind, pwr_loss, voltage_loss, max_current, net_names, length = traceinfo(CU_THICK, True)

                window_sizer = wx.BoxSizer()
                window_sizer.Add(panel, 1, wx.ALL | wx.EXPAND)

                sizer = wx.GridBagSizer(10, 0)
                panel.SetBackgroundColour("white")
                #sizer = wx.BoxSizer(wx.VERTICAL)
                sizer.Add(wx.StaticText(panel, label="Net#"), border=4, pos=(0, 0))
                sizer.Add(wx.StaticText(panel, label="Name"), border=4, pos=(0, 1))
                sizer.Add(wx.StaticText(panel, label="Length [mm]"), border=4, pos=(0, 2))
                sizer.Add(wx.StaticText(panel, label="Resistance [Ohm]"), border=4, pos=(0, 3))
                sizer.Add(wx.StaticText(panel, label="Inductance [H]"), border=4, pos=(0, 4))
                sizer.Add(wx.StaticText(panel, label="Power loss [W]"), border=4, pos=(0, 5))
                sizer.Add(wx.StaticText(panel, label="Voltage loss [V]"), border=4, pos=(0, 6))
                sizer.Add(wx.StaticText(panel, label="Max Current [A]"), border=4, pos=(0, 7))
                self.lb_net = {}
                self.lb_name = {}
                self.lb_res = {}
                self.lb_ind = {}
                self.lb_pwr = {}
                self.lb_vdo = {}
                self.lb_max = {}
                self.lb_len = {}
                self.row = {}
                row = 1
                for net in res:
                    self.lb_net[row] = wx.StaticText(panel, label=str(net))
                    self.lb_name[row] = wx.StaticText(panel, label=str(net_names[net]))
                    self.lb_len[row] = wx.StaticText(panel, label=format_number(length[net]))
                    self.lb_res[row] = wx.StaticText(panel, label=format_number(res[net]))
                    self.lb_ind[row] = wx.StaticText(panel, label=format_number(ind[net]))
                    self.lb_pwr[row] = wx.StaticText(panel, label=format_number(pwr_loss[net]))
                    self.lb_vdo[row] = wx.StaticText(panel, label=format_number(voltage_loss[net]))
                    self.lb_max[row] = wx.StaticText(panel, label=format_number(max_current[net]))
                    self.lb_net[row].SetBackgroundColour('#cfcfcf')
                    #lb_max[row] = wx.StaticText(panel, label=str("%.2e"%([net]))

                    sizer.Add(self.lb_net[row], flag=wx.ALL|wx.EXPAND, border=4, pos=(row, 0))
                    sizer.Add(self.lb_name[row], flag=wx.ALL|wx.EXPAND, border=4, pos=(row, 1))
                    sizer.Add(self.lb_len[row], flag=wx.ALL|wx.EXPAND, border=4, pos=(row, 2))
                    sizer.Add(self.lb_res[row], flag=wx.ALL|wx.EXPAND, border=4, pos=(row, 3))
                    sizer.Add(self.lb_ind[row], flag=wx.ALL|wx.EXPAND, border=4, pos=(row, 4))
                    sizer.Add(self.lb_pwr[row], flag=wx.ALL|wx.EXPAND, border=4, pos=(row, 5))
                    sizer.Add(self.lb_vdo[row], flag=wx.ALL|wx.EXPAND, border=4, pos=(row, 6))
                    sizer.Add(self.lb_max[row], flag=wx.ALL|wx.EXPAND, border=4, pos=(row, 7))
                    self.row[net] = row
                    row += 1

                bottom = wx.BoxSizer(wx.HORIZONTAL)

                self.thickradio = wx.RadioBox(panel, wx.ID_ANY, label="Thickness [um]", choices=['18', '35', '70'])
                self.thickradio.Bind(wx.EVT_RADIOBOX, self.update_display)
                bottom.Add(self.thickradio, 0, wx.RIGHT, 4) #, (row, 1))
                self.intext = wx.RadioBox(panel, wx.ID_ANY, label="Internal/External", choices=['int', 'ext'])
                self.intext.Bind(wx.EVT_RADIOBOX, self.update_display)
                bottom.Add(self.intext, 0, wx.RIGHT, 4)
                # ------------------
                close_button = wx.Button(panel, label="Close")
                close_button.Bind(wx.EVT_BUTTON, self.on_button_close)
                bottom.Add(close_button, 0, wx.LEFT, 4) #, (row, 6))
                self.maxrows = row

                border = wx.BoxSizer(wx.VERTICAL)
                border.Add(sizer, 3, wx.ALL | wx.EXPAND, 5)
                border.Add(bottom, 3, wx.TOP | wx.EXPAND, 5)
                panel.SetSizerAndFit(border)
                self.SetSizerAndFit(window_sizer)
                #self.Fit()

            def update_display(self, event):
                event.Skip()
                if self.thickradio.GetSelection() == 0:
                    thickness = 18e-6
                elif self.thickradio.GetSelection() == 1:
                    thickness = 35e-6
                elif self.thickradio.GetSelection() == 2:
                    thickness = 70e-6
                if self.intext.GetSelection() == 0:
                    internal = True
                else:
                    internal = False
                res, ind, pwr_loss, voltage_loss, max_current, net_names, length = traceinfo(thickness, internal)
                for net in res:
                    row = self.row[net]
                    self.lb_net[row].SetLabel(str(net))
                    self.lb_name[row].SetLabel(str(net_names[net]))
                    self.lb_len[row].SetLabel(format_number(length[net]))
                    self.lb_res[row].SetLabel(format_number(res[net]))
                    self.lb_ind[row].SetLabel(format_number(ind[net]))
                    self.lb_pwr[row].SetLabel(format_number(pwr_loss[net]))
                    self.lb_vdo[row].SetLabel(format_number(voltage_loss[net]))
                    self.lb_max[row].SetLabel(format_number(max_current[net]))


            def on_button_close(self, event):
                event.Skip()
                self.Close()
        frame = DisplayResults(None)
        frame.Center()
        frame.ShowModal()
        frame.Destroy()
